### PR TITLE
# Inference `devshards`

### PR DESCRIPTION
**Pull request by gmorgachev**
# Inference shards

Per-inference transactions on mainnet limit throughput to ~277 inferences/sec. Requests could be batched, but computation and state growth per request make this unscalable to hundreds of thousands of inferences.

This PR moves all per-inference communication off-chain. The chain processes only two transactions per session: one to put coins in escrow and assign a subgroup of hosts, one to settle at the end. All inference requests and validations happen inside the subgroup directly, over a long session.

To close the session, the user submits the final usage state signed by a supermajority of hosts (threshold: 2/3 slot-weighted). Both sides have a clear incentive to settle -- the user recovers unused escrow balance, the subgroup gets paid from it.

As each subgroup achieves consensus for the final state, the architecture becomes:
- main blockchain
- many sub-chains / shards with extremely lightweight architecture

```
+-----------+     +-------------------+     +----------------------------+
|   User    |     |      Mainnet      |     |  Subnet (one per session)  |
+-----------+     +-------------------+     +----------------------------+
      |                    |                             |
      | 1. MsgCreateEscrow |                             |
      |    (100GNK)        |                             |
      | -----------------> |                             |
      | <- escrow_id,      |                             |
      |    group=[h1..hN]  |                             |
      |                    |                             |
      | 2. POST /chat (req1) --------------------------> |
      | 3. POST /chat (req2) --------------------------> |
      | 4. POST /chat (reqN) --------------------------> |
      |    ...             |                             |
      |                    |                             |
      | 5. MsgSettleEscrow |                             |
      |   (finalState,     |                             |
      |    signatures, ..) |                             |
      | -----------------> |                             |
      | <- user refund +   |                             |
      |    hosts paid      |                             |
+-----------+     +-------------------+     +----------------------------+
```

The user acts as the sequencer. They order transactions, propagate state to hosts, and initiate any timeout or settlement procedures. The user is directly incentivized to do this -- unresolved inferences lock escrow balance, and unrecorded host work reduces the user's refund at settlement.

## Implementation

This is version 0 of subnets. The protocol will evolve, but it is important to start testing on a real chain now. Early mainnet testing will surface issues that unit tests and local testnets cannot.

We created a separate standalone module `/subnet/` with no Cosmos SDK dependencies. This boundary exists so the subnet code can be tested and reasoned about independently. Long-term goal is to fully separate the subnet service from `decentralized-api` into its own process.

In this early version, the subnet module is compiled into `decentralized-api` and runs in the same process. This avoids the operational overhead of a separate deployment while the protocol stabilizes.

### What lives where

`/subnet/` contains the core protocol logic: state machine that applies diffs, tracks inference lifecycle through 7 states (pending, started, finished, challenged, validated, invalidated, timed_out), computes costs, validates secp256k1 signatures (user, executor, host-proposed), and derives executor slots deterministically from nonce (`inference_id % group_size`). Includes HTTP transport via Echo, gossip with K-fanout nonce propagation, and SQLite storage (WAL mode) for append-only diff logs. Uses `go-ethereum/crypto` for signing, `google.golang.org/protobuf` for deterministic serialization. No shared types with mainnet protos.

The subnet defines 8 off-chain transaction types: `MsgStartInference`, `MsgConfirmStart`, `MsgFinishInference`, `MsgTimeoutInference`, `MsgValidation`, `MsgValidationVote`, `MsgRevealSeed`, `MsgFinalizeRound`.

`decentralized-api/internal/subnet/` adapts the subnet module to dapi. `EngineAdapter` wraps dapi's broker and completionapi to execute inferences on vLLM nodes. `ValidationAdapter` re-executes inferences with enforced tokens and compares logits. `ChainBridge` implements `MainnetBridge` via gRPC for escrow queries and warm key verification. `HostManager` creates sessions lazily per escrow, recovers them from SQLite on restart, and registers endpoints on `/subnet/v1/`.

`inference-chain/x/inference/keeper/` adds two new chain messages. `MsgCreateSubnetEscrow` locks funds and samples 16 slots via weighted random sampling using `GetSlotsFromSorted` (same algorithm as PoC validation). `MsgSettleSubnetEscrow` verifies the settlement proof -- recomputes `host_stats_hash`, checks `SHA256(host_stats_hash || rest_hash || phase_byte) == state_root`, verifies 2/3+1 slot-weighted secp256k1 signatures, distributes payments to validators proportionally to accumulated cost, refunds the remainder. Unsettled escrows older than 2 epochs are pruned; funds split equally among unique validators.

### Protocol properties

Escrow accounting is checked per-inference. `MsgStartInference` reserves `(input_length + max_tokens) * token_price` from the escrow balance. `MsgFinishInference` computes actual cost from real token counts and releases the surplus. `MsgTimeoutInference` releases the full reservation.

Hosts measure executor performance. Each slot accumulates missed count (timeouts), invalid count (failed validations), and total cost in `host_stats`. These stats are part of the settlement proof.

Validation is probabilistic. Each host derives a deterministic seed by signing the escrow ID bytes and taking the first 8 bytes. `ShouldValidate` computes a per-inference probability from the seed and a configurable rate. Seeds are revealed during a mandatory finalizing round before settlement. Compliance (required vs completed validations) is computed deterministically from revealed seeds.

Warm keys are supported. Hosts can sign with operator keys authorized via on-chain authz grants. The first diff-contained signature from a slot binds the warm key permanently for the session. Warm key bindings are part of the state root.

Timeout verification requires signed votes from other hosts as evidence. Two reasons: `refused` (executor never signed receipt, pending -> timed_out) and `execution` (executor signed but didn't finish within deadline, started -> timed_out). Threshold: total_slots/2.

Equivocation (user submitting conflicting diffs at the same nonce) is detectable via gossip and results in session termination.

State root is `SHA256(host_stats_hash || rest_hash || phase_byte)`, where `rest_hash = SHA256(balance || inferences_hash || warm_keys_hash)`. All hashing uses deterministic proto serialization.

## Next steps

This PR proposes merging subnets into v0.2.11 with limited access. `AllowedCreatorAddresses` in `SubnetEscrowParams` restricts who can create escrows. The plan is to test on mainnet, resolve issues, and remove the restriction in v0.2.12-v0.2.13.

The subnet module is experimental. All parameters are conservative defaults designed to avoid harm to the chain: group size 16 (slots per escrow), escrow bounds 5-10 GNK, max 100 escrows per epoch, unsettled escrows pruned after 2 epochs.

Issues to resolve before wider rollout:

- Aggregating subnet stats on mainnet. Subnet `host_stats` (missed, invalid, cost, validations) do not yet feed into the bitcoin-style block reward or PoC reputation system.
- Skeptical review of the state machine. Adversarial edge cases (e.g. user-host collusion, strategic timeout abuse) need formal analysis.
- Timeout flows end-to-end. Refusal and execution timeout verification need integration testing under real network conditions.
- Pricing. Token price is flat per session, we need to adapt dynamic
- BLS signature aggregation. `MsgSettleSubnetEscrow` currently carries N individual secp256k1 signatures. Aggregating into a single BLS signature reduces transaction size.
- ...

## References

- [README](proposals/inference/README.md)
- [Design](proposals/inference/design.md)
- [Implementation plan](proposals/inference/implementation.md)

### CODE REVIEW

File: decentralized-api/internal/subnet/signer.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+func NewSignerFromKeyring(kr keyring.Keyring, uid string) (*signing.Secp256k1Signer, error) {
+	armor, err := kr.ExportPrivKeyArmor(uid, "")
+	if err != nil {
+		return nil, fmt.Errorf("export priv key: %w", err)
+	}
+
+	privKey, _, err := sdkcrypto.UnarmorDecryptPrivKey(armor, "")
+	if err != nil {
+		return nil, fmt.Errorf("unarmor priv key: %w", err)
+	}
```
[the remark this answers is not in the document]
Comment (gmorgachev) [reply decentralized-api/internal/subnet/signer.go:24]: yes, should be fixed

**Comment by gmorgachev** [comment]
# TODO: 
- how subnet understand that escrow expired
- no float point in state machine
- timeout during POC
- BLS
- prunning of on-chain / offchain data
- tests for both timeouts flow 
- timeout before settlement 
- tests for refusing to sign if TX is not recorded
- DDOS protection for endpoints
- base payment per escrow

File: inference-chain/x/inference/keeper/msg_server_create_subnet_escrow.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+func (k msgServer) CreateSubnetEscrow(goCtx context.Context, msg *types.MsgCreateSubnetEscrow) (*types.MsgCreateSubnetEscrowResponse, error) {
+	if err := k.CheckPermission(goCtx, msg, NoPermission); err != nil {
+		return nil, err
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	ep := k.GetSubnetEscrowParams(goCtx)
+
+	if !k.IsAllowedEscrowCreator(goCtx, msg.Creator) {
```
Comment (patimen) [inline inference-chain/x/inference/keeper/msg_server_create_subnet_escrow.go]: This needs to be moved to CheckPermission as a proper Permission... that's the entire point of having CheckPermission.

**Review by patimen — APPROVED** [review]
I've reviewed and fixed issues that might impact normal chain operation.
I have not significantly reviewed the direct implementations of the new subnet feature. I may come up with some issues on that going forward, but given how it is protected by an allowlist I am comfortable with that.
